tests/ApplySecureHeadersTest.php

```diff
@@ -93,45 +94,114 @@ public function testMiddlewareAddsExpectCTHeaders()
+        ];
+
+        $response = $this->applySecureHeadersWithConfig(new Response(), $config);
+        $headers = $response->headers->all();
+
+        $this->setRequest($response);
+
+        $this->assertContains($config['cspNonces'][0], $headers['content-security-policy'][0]);
+        $this->assertContains('nonce-', csp_nonce($config['cspNonces'][0]));
+    }
+
+    /**
+     * Ensure an exception is thrown if the csp_nonce function is used without setting the header.
+     *
+     * @return void
+     * @throws ContentSecurityPolicyNotFoundException
+     */
+    public function testCspNonceFunctionFailsWhenNotSet()
+    {
+        $this->expectException(ContentSecurityPolicyNotFoundException::class);
+
+        $this->setRequest(new Response());
+
+        csp_nonce('foo');
+    }
+
     /**
      * Apply SecureHeaders from the given config to a Response.
      *
      * @param Response $response
-     * @param string|null $key
-     * @param array|null $configMap
+     * @param array $configMap
      * @return Response
      */
-    private function applySecureHeadersWithConfig(Response $response, string $key = null, array $configMap = null)
+    private function applySecureHeadersWithConfig(Response $response, array $configMap = [])
     {
+        // convert the configMap dot keys into deeply nested array
+        $configMap = [
+            'secure-headers' => ! empty($configMap)
+                ? array_merge_recursive(...array_map(
+                    [self::class, 'nonMutatingDataFill'],
+                    array_fill(0, count($configMap), []),
+                    array_keys($configMap),
+                    $configMap
+                ))
+                : []
+        ];
+        
         /** @var Repository|\PHPUnit_Framework_MockObject_MockObject $config */
         $config = $this->createMock(Repository::class);
-
         $config->expects($this->any())
             ->method('get')
             ->with($this->anything())
-            ->will($this->returnCallback(function () use ($key, $configMap) {
-                $args = func_get_args();
-
-                if ($args[0] === "secure-headers.{$key}") {
-                    return $configMap;
-                }
-
-                return $args[1];
+            ->will($this->returnCallback(function (string $key, $default) use ($configMap) {
+                return array_get($configMap, $key, $default);
             }));
 
         $secureHeaders = new SecureHeaders();
         $secureHeaders->errorReporting(false);
+
         $middleware = new ApplySecureHeaders($config, $secureHeaders);
 
         return $middleware->handle(new Request, function () use ($response) { return $response; });
     }
+
+    /**
+     * Fill an array without mutating the data.
+     *
+     * @param array $array
+     * @param string $key
+     * @param $value
+     * @return array
+     */
+    private static function nonMutatingDataFill(array $array, string $key, $value) {
```

Comment:
Possibly these changes (and the other related ones) can be added to my PR to save creating merge conflicts? ;p

If you're happy with my PR in it's current state then once it is merged then this PR can rebased to pick the new test methods up, or otherwise I'd maybe suggest that I split the test adjustments from my feature additions in #15 so that the new tests can be merged in ahead so this PR can rely on them?

Reply:
Cool - I have merged yours as yeah it was where I grabbed those beautiful test methods from. Will rebase 👍